Pwndbg architecture name list

OBarronCS opened the pull request:
This PR creates a single "point-of-truth" list for the names of architectures that Pwndbg supports.

See comments for rationale on specific changes.

OBarronCS committed:
Create an authoritative list of architectures that Pwndbg supports. Add test to ensure some important mappings exist - add i8086 pwnlib mapping

OBarronCS commented on pwndbg/aglib/arch.py:
```diff
@@ -5,41 +5,9 @@
 from pwndbg.lib.arch import Arch
 
-# List of architectures - used when determining GDB arch, and in OnlyWithArch
-ARCHS = (
-    "x86-64",
-    "i386",
-    "aarch64",
-    "mips",
-    "powerpc",
-    "sparc",
-    "arm",
-    "armcm",
-    "riscv:rv32",
-    "riscv:rv64",
-    "riscv",
-    "loongarch64",
-)
```
These names are specific to GDB (see `riscv:32`, `riscv:64`), so it was moved to the GDB specific file.

OBarronCS commented on pwndbg/dbg/gdb/__init__.py:
```diff
@@ -35,6 +35,25 @@
+gdb_architecture_name_fixup_list = (
+    "x86-64",
+    "i386",
+    "i8086",
```
This `i8086` mapping is new. On the early stages of an x86 bootloader, the first ~20 instructions are in i8086 mode before the CPU converts to a 32-bit mode. Now Pwndbg doesn't crash in these cases.

OBarronCS commented on pwndbg/lib/arch.py:
```diff
@@ -2,8 +2,49 @@
 
+# Names of arches that Pwndbg supports
+# The names that GDB/LLDB recognize are converted to a name in this list
+PWNDBG_SUPPORTED_ARCHITECTURES_TYPE = Literal[
+    "x86-64",
+    "i386",
+    "i8086",
+    "mips",
+    "aarch64",
+    "arm",
+    "armcm",
+    "iwmmxt",
+    "rv32",
+    "rv64",
+    "sparc",
+    "powerpc",
+    "loongarch64",
+]
+
+PWNDBG_SUPPORTED_ARCHITECTURES: list[PWNDBG_SUPPORTED_ARCHITECTURES_TYPE] = list(
+    typing.get_args(PWNDBG_SUPPORTED_ARCHITECTURES_TYPE)
+)
```
The `Literal` above will provide "type-safety" (red underline in IDE) / auto-complete on the architecture name. Then the elements are extracted into a list that can be used during runtime.

disconnect3d replied:
Omg I need to revisit this as I don't understand it :D

OBarronCS commented on pwndbg/lib/arch.py:
```diff
@@ -2,8 +2,49 @@
+PWNLIB_ARCH_MAPPINGS = {
+    "x86-64": "amd64",
+    "i386": "i386",
+    "i8086": "none",
```
New mapping to prevent crash in i8086 mode.

patryk4815 replied:
do we need this mapping at all? maybe lets use it as:
`PWNLIB_ARCH_MAPPINGS.get(archname, "none")` ?

OBarronCS replied:
This is a good call - I made this change.

OBarronCS committed:
make pwnlib mapping 'none' by default

patryk4815 commented on pwndbg/dbg/gdb/__init__.py:
```diff
@@ -35,6 +35,25 @@
 
+# List of supported architectures that GDB recognizes
+# These strings to converted to the Pwndbg-specific name for the architecture
+gdb_architecture_name_fixup_list = (
+    "x86-64",
+    "i386",
+    "i8086",
+    "aarch64",
+    "mips",
+    "powerpc",
+    "sparc",
+    "arm",
+    "armcm",
+    "riscv:rv32",
+    "riscv:rv64",
+    "riscv",
+    "loongarch64",
+)
+
+
```
```suggestion
gdb_architecture_name_fixup_list = (
    "x86-64",
    "i386",
    "i8086",
    "aarch64",
    "mips",
    "powerpc",
    "sparc",
    "arm",
    "riscv:rv32",
    "riscv:rv64",
    "riscv",
    "loongarch64",
)
```
removed `armcm`, gdb don't have this arch type

patryk4815 commented on pwndbg/lib/arch.py:
```diff
@@ -2,8 +2,49 @@
+    "i8086": "none",
+    "mips": "mips",
+    "aarch64": "aarch64",
+    "arm": "arm",
+    "iwmmxt": "arm",
+    "armcm": "thumb",
```
```suggestion
    "arm": "arm",
    "armcm": "thumb",
```
removed `iwmmxt` (probably it is already handled by `arm` name)

patryk4815 commented on pwndbg/lib/arch.py:
```diff
@@ -2,8 +2,49 @@
+    "mips",
+    "aarch64",
+    "arm",
+    "armcm",
+    "iwmmxt",
+    "rv32",
```
```suggestion
    "armcm",
    "rv32",
```
removed `iwmmxt` (probably it is already handled by `arm` name)

patryk4815 commented on pwndbg/lib/arch.py:
```diff
@@ -2,8 +2,49 @@
+    "loongarch64",
+]
+
+PWNDBG_SUPPORTED_ARCHITECTURES: list[PWNDBG_SUPPORTED_ARCHITECTURES_TYPE] = list(
```
Can we add also this type checking here?
https://github.com/pwndbg/pwndbg/blob/e99d4e7f0b9419c3929bceabe0a624faac4866ab/pwndbg/lib/regs.py#L673
And remove `iwmmxt` register sets

OBarronCS replied:
Done!

patryk4815 commented:
Only ARM architecture in GDB is handled differently then other architectures (others have proper prefix eg: "i386:x86-64", "riscv:rv64").  
So `xscale`, `iwmmxt`, and `iwmmxt2` must be manually mapped to `arm`.
<img width="655" alt="image" src="https://github.com/user-attachments/assets/4ba6b1ba-9796-49f8-b582-f2180acecf2f" />

Or we have to replace `show architecture` with this func:
```
In [3]: gdb.selected_inferior().architecture()
Out[3]: <gdb.Architecture arch_name=arm printable_name=armv8-m.main>
# Or if proc is alive:
In [4]: gdb.newest_frame().architecture()
```
(sic!) only repr show `arch_name=arm`, so we have to parse repr manually

disconnect3d commented:
Ping @OBarronCS  see comments

OBarronCS committed:
Update pwndbg/dbg/gdb/__init__.py

Co-authored-by: patryk4815 <bux.patryk@gmail.com>

OBarronCS committed:
Update pwndbg/lib/arch.py

Co-authored-by: patryk4815 <bux.patryk@gmail.com>

OBarronCS committed:
Update pwndbg/lib/arch.py

Co-authored-by: patryk4815 <bux.patryk@gmail.com>

OBarronCS committed:
More type safety with the list of supported types, handle iwmmxt, iwmmxt2, and xscale manually

OBarronCS commented:
I removed the definitions of `iwmmxt` (which were just duplicates of `arm`) as noted in the above comments, and I added an `if` check inside the GDB arch detection function to map `iwmmxt`,`iwmmxt2`, and `xscale` to `arm`. I also added the new support arch type literal to a couple places for more mypy type safety.
